master/internal/stream/testutils.go

```diff
@@ -0,0 +1,234 @@
+func (s *mockSocket) ReadIncoming(data *interface{}) error {
+	select {
+	case <-s.closed:
+		return &websocket.CloseError{Code: websocket.CloseAbnormalClosure}
+	case msg := <-s.inbound:
+		*data = msg
+		return nil
+	}
+}
+
+// ReadUntil reads until the terminationMsg has been read.
+func (s *mockSocket) ReadUntil(data *[]interface{}, terminationMsg interface{}) error {
+	var msg interface{}
+	for {
+		if reflect.TypeOf(msg) == reflect.TypeOf(terminationMsg) {
+			switch typedMsg := msg.(type) {
+			case stream.UpsertMsg:
+				if err := validateUpsertMsg(typedMsg.Msg, terminationMsg.(stream.UpsertMsg).Msg); err == nil {
+					return nil
+				}
+			case string:
+				if typedMsg == terminationMsg.(string) {
+					return nil
+				}
+			case SyncMsg:
+				if typedMsg.SyncID == terminationMsg.(SyncMsg).SyncID {
+					return nil
+				}
+			}
+		}
+		if err := s.ReadIncoming(&msg); err != nil {
+			return err
+		}
+		*data = append(*data, msg)
+	}
+}
+
+// Close closes the mockSocket.
+func (s *mockSocket) Close() {
+	close(s.closed)
+}
+
+func splitMsgs[M stream.Msg](messages []interface{}) (
+	deletions []string,
+	upserts []stream.Msg,
+	syncs []SyncMsg,
+	err error,
+) {
+	typeHolder := new(M)
+	for _, msg := range messages {
+		if deletion, ok := msg.(stream.DeleteMsg); ok {
+			deletions = append(deletions, deletion.Deleted)
+		} else if upsert, ok := msg.(stream.UpsertMsg); ok {
+			upsertM, ok := upsert.Msg.(M)
+			if !ok {
+				return nil, nil, nil, fmt.Errorf("expected %T, but received %T", *typeHolder, upsert.Msg)
+			}
+			upserts = append(upserts, upsertM)
+		} else if syncMsg, ok := msg.(SyncMsg); ok {
+			syncs = append(syncs, syncMsg)
+		} else {
+			return nil, nil, nil, fmt.Errorf("expected a string or %T, but received %T", *typeHolder,
+				reflect.TypeOf(msg).Name())
+		}
+	}
+	return deletions, upserts, syncs, nil
+}
+
+func validateMsgs[M stream.Msg](
+	sync SyncMsg,
+	expectedSync SyncMsg,
+	upserts []stream.Msg,
+	expectedUpserts []M,
+	deletions []string,
+	expectedDeletions []string,
+) error {
+	switch {
+	// check if we received the correct number of trial messages
+	case len(upserts) != len(expectedUpserts):
+		return fmt.Errorf(
+			"did not receive expected number of upsert messages: expected %d, actual: %d",
+			len(expectedUpserts),
+			len(upserts),
+		)
+	// check if we received the correct number of deletion messages
+	case len(deletions) != len(expectedDeletions):
+		return fmt.Errorf(
+			"did not receive expected number of deletion messages: expected %v, actual: %v",
+			len(expectedDeletions),
+			len(deletions),
+		)
+	// check if content of messages is correct
+	default:
+		if sync.SyncID != expectedSync.SyncID {
+			return fmt.Errorf(
+				"did not receive expected sync message: expected: %v, actual: %v",
+				expectedSync,
+				sync,
+			)
+		}
+		// XXX: this expects messages to be sent in a deterministic order, is this actually enforced?
```

Comment:
Sounds good, I actually really like the map approach; if we have it so the test case provides a [TrialMsg]struct{} + since we are guaranteed that we have the right number of messages; so we just have to make sure that none of the actual messages don't have a entry in the map

Reply:
So we can't directly go to using the map approach for upserts since `Msg` doesn't implement `Comparable`; however, if we use something like `msg.testString()`, it makes things significantly easier to validate upsertMsgs because it just becomes a string comparison + the test owns on how the strings are created.

This also simplifies the test case struct, removing the need to be typed to a specific streamble type, as well. This also means we could have something like, startup tests for every type could be a single test (maybe using subtests instead of error groups)